File: src/data/drcLocations.ts
// Villes de la RDC et leurs communes officielles — champs "Ville"/"Commune" du formulaire papier.
// Source : Wikipédia (listes des villes/communes de la RDC, structure provinciale 2015) + jeu de
// données DRC.Geo (github.com/Jonathan-S24/drc-geo) pour la liste des villes par province.
// Liste non garantie exhaustive à 100% (des villes existent sans liste de communes vérifiée
// ci-dessous) — d'où l'option "Autre" en secours dans les deux champs.
export interface VilleEntry {
  name: string
  province: string
  communes: string[]
}

export const DRC_VILLES: VilleEntry[] = [
  { name: 'Kinshasa', province: 'Kinshasa', communes: [
    'Bandalungwa', 'Barumbu', 'Bumbu', 'Gombe', 'Kalamu', 'Kasa-Vubu', 'Kimbanseke', 'Kinshasa',
    'Kintambo', 'Kisenso', 'Lemba', 'Limete', 'Lingwala', 'Makala', 'Maluku', 'Masina', 'Matete',
    'Mont-Ngafula', 'Ndjili', 'Ngaba', 'Ngaliema', 'Ngiri-Ngiri', 'Nsele', 'Selembao',
  ] },
  { name: 'Buta', province: 'Bas-Uele', communes: ['Babade', 'Dobea', 'Finant', 'Tepatondele'] },
  { name: 'Aketi', province: 'Bas-Uele', communes: [] },
  { name: 'Dingila', province: 'Bas-Uele', communes: [] },
  { name: 'Mbandaka', province: 'Équateur', communes: ['Mbandaka', 'Wangata'] },
  { name: 'Likasi', province: 'Haut-Katanga', communes: ['Kikula', 'Likasi', 'Panda', 'Shituru'] },
  { name: 'Lubumbashi', province: 'Haut-Katanga', communes: [
    'Annexe', 'Kamalondo', 'Kampemba', 'Katuba', 'Kenya', 'Lubumbashi', 'Ruashi',
  ] },
  { name: 'Kipushi', province: 'Haut-Katanga', communes: [] },
  { name: 'Kamina', province: 'Haut-Lomami', communes: ['Dimayi', 'Kamina', 'Sobongo', 'Kupa'] },
  { name: 'Isiro', province: 'Haut-Uele', communes: ['Mambaya', 'Mendambo', 'Mbunya'] },
  { name: 'Bunia', province: 'Ituri', communes: ['Nyakasanza', 'Mbunya', 'Shari'] },
  { name: 'Ariwara', province: 'Ituri', communes: [] },
  { name: 'Mongwalu', province: 'Ituri', communes: [] },
  { name: 'Tshikapa', province: 'Kasaï', communes: ['Dibumba I', 'Dibumba II', 'Kanzala', 'Mabondo', 'Mbumba'] },
  { name: 'Kananga', province: 'Kasaï-Central', communes: ['Kananga', 'Katoka', 'Lukonga', 'Ndesha', 'Nganza'] },
  { name: 'Tshimbulu', province: 'Kasaï-Central', communes: [] },
  { name: 'Mbuji-Mayi', province: 'Kasaï-Oriental', communes: ['Bipemba', 'Dibindi', 'Diulu', 'Kanshi', 'Muya'] },
  { name: 'Boma', province: 'Kongo-Central', communes: ['Kabondo', 'Kalamu', 'Nzadi'] },
  { name: 'Matadi', province: 'Kongo-Central', communes: ['Matadi', 'Mvuzi', 'Nzanza'] },
  { name: 'Mbanza-Ngungu', province: 'Kongo-Central', communes: [] },
  { name: 'Inkisi', province: 'Kongo-Central', communes: [] },
  { name: 'Kenge', province: 'Kwango', communes: ['Cinq-Mai', 'Laurent-Désiré-Kabila', 'Manonga', 'Masikita', 'Mavula'] },
  { name: 'Bandundu', province: 'Kwilu', communes: ['Basoko', 'Disasi', 'Mayoyo'] },
  { name: 'Kikwit', province: 'Kwilu', communes: ['Kazamba', 'Kikwit', 'Lukemi', 'Lukolela', 'Nzinda'] },
  { name: 'Kabinda', province: 'Lomami', communes: ['Kabondo', 'Kabuela-Buela', 'Kajiba', 'Mudingayi'] },
  { name: 'Mwene-Ditu', province: 'Lomami', communes: ['Bondoyi', 'Musadi', 'Mwene-Ditu'] },
  { name: 'Kolwezi', province: 'Lualaba', communes: ['Dilala', 'Kolwezi', 'Manika'] },
  { name: 'Kasaji', province: 'Lualaba', communes: [] },
  { name: 'Inongo', province: 'Mai-Ndombe', communes: ['Bonse', 'Inongo', 'Mpolo', 'Mpongonzoli'] },
  { name: 'Nioki', province: 'Mai-Ndombe', communes: [] },
  { name: 'Kindu', province: 'Maniema', communes: ['Alunguli', 'Kasuku', 'Kindu', 'Mikelenge'] },
  { name: 'Kalima', province: 'Maniema', communes: [] },
  { name: 'Lisala', province: 'Mongala', communes: ['Bolikango', 'Lisala', 'Mongala'] },
  { name: 'Bumba', province: 'Mongala', communes: [] },
  { name: 'Gbadolite', province: 'Nord-Ubangi', communes: ['Gbadolite', 'Molegbe', 'Nganza'] },
  { name: 'Beni', province: 'Nord-Kivu', communes: ['Beu', 'Bungulu', 'Mulekera', 'Ruwenzori'] },
  { name: 'Butembo', province: 'Nord-Kivu', communes: ['Bulengera', 'Butembo', 'Kimemi', 'Mususa', 'Vulamba'] },
  { name: 'Goma', province: 'Nord-Kivu', communes: ['Goma', 'Karisimbi'] },
  { name: 'Lusambo', province: 'Sankuru', communes: ['Kabondo', 'Lupembe', 'Lusambo', 'Tusuanganyi'] },
  { name: 'Wembo-Nyama', province: 'Sankuru', communes: ['Ewango', 'Lumumbaville', 'Wembo-Nyama'] },
  { name: 'Bukavu', province: 'Sud-Kivu', communes: ['Bagira', 'Ibanda', 'Kadutu'] },
  { name: 'Baraka', province: 'Sud-Kivu', communes: [] },
  { name: 'Kamituga', province: 'Sud-Kivu', communes: [] },
  { name: 'Uvira', province: 'Sud-Kivu', communes: [] },
  { name: 'Gemena', province: 'Sud-Ubangi', communes: ['Gbazubu', 'Labo', 'Lac-Ntumba', 'Mont-Gila'] },
  { name: 'Zongo', province: 'Sud-Ubangi', communes: [] },
  { name: 'Kalemie', province: 'Tanganyika', communes: ['Kalemie', 'Lac', 'Lukuga'] },
  { name: 'Kisangani', province: 'Tshopo', communes: ['Kabondo', 'Kisangani', 'Lubunga', 'Makiso', 'Mangobo', 'Tshopo'] },
  { name: 'Yangambi', province: 'Tshopo', communes: [] },
  { name: 'Boende', province: 'Tshuapa', communes: ['Boende', 'Tshuapa'] },
]

export const AUTRE_VILLE = 'AUTRE'

export function communesForVille(ville: string): string[] {
  return DRC_VILLES.find((v) => v.name.toUpperCase() === ville.toUpperCase())?.communes ?? []
}

// Les 26 provinces de la RDC — champ "Province" du formulaire, placé avant la ville.
export const DRC_PROVINCES: string[] = [
  'Bas-Uele', 'Équateur', 'Haut-Katanga', 'Haut-Lomami', 'Haut-Uele', 'Ituri', 'Kasaï',
  'Kasaï-Central', 'Kasaï-Oriental', 'Kinshasa', 'Kongo-Central', 'Kwango', 'Kwilu', 'Lomami',
  'Lualaba', 'Mai-Ndombe', 'Maniema', 'Mongala', 'Nord-Kivu', 'Nord-Ubangi', 'Sankuru',
  'Sud-Kivu', 'Sud-Ubangi', 'Tanganyika', 'Tshopo', 'Tshuapa',
]

/** Province (en majuscules, comme le reste de la saisie) d'une ville de la liste, ou '' pour une ville saisie librement. */
export function provinceForVille(ville: string): string {
  return DRC_VILLES.find((v) => v.name.toUpperCase() === ville.toUpperCase())?.province.toUpperCase() ?? ''
}

File: android-app/app/src/main/java/com/onip/cartoonip/data/DrcLocations.kt
package com.onip.cartoonip.data

// Villes de la RDC et leurs communes officielles — champs "Ville"/"Commune" du formulaire papier.
// Même source et mêmes limites que la liste équivalente côté web (src/data/drcLocations.ts) :
// non garantie exhaustive à 100%, d'où l'option "Autre" en secours dans les deux champs.
data class VilleEntry(val name: String, val province: String, val communes: List<String>)

const val AUTRE_VILLE = "AUTRE"

val DRC_VILLES: List<VilleEntry> = listOf(
    VilleEntry("Kinshasa", "Kinshasa", listOf(
        "Bandalungwa", "Barumbu", "Bumbu", "Gombe", "Kalamu", "Kasa-Vubu", "Kimbanseke", "Kinshasa",
        "Kintambo", "Kisenso", "Lemba", "Limete", "Lingwala", "Makala", "Maluku", "Masina", "Matete",
        "Mont-Ngafula", "Ndjili", "Ngaba", "Ngaliema", "Ngiri-Ngiri", "Nsele", "Selembao",
    )),
    VilleEntry("Buta", "Bas-Uele", listOf("Babade", "Dobea", "Finant", "Tepatondele")),
    VilleEntry("Aketi", "Bas-Uele", emptyList()),
    VilleEntry("Dingila", "Bas-Uele", emptyList()),
    VilleEntry("Mbandaka", "Équateur", listOf("Mbandaka", "Wangata")),
    VilleEntry("Likasi", "Haut-Katanga", listOf("Kikula", "Likasi", "Panda", "Shituru")),
    VilleEntry("Lubumbashi", "Haut-Katanga", listOf(
        "Annexe", "Kamalondo", "Kampemba", "Katuba", "Kenya", "Lubumbashi", "Ruashi",
    )),
    VilleEntry("Kipushi", "Haut-Katanga", emptyList()),
    VilleEntry("Kamina", "Haut-Lomami", listOf("Dimayi", "Kamina", "Sobongo", "Kupa")),
    VilleEntry("Isiro", "Haut-Uele", listOf("Mambaya", "Mendambo", "Mbunya")),
    VilleEntry("Bunia", "Ituri", listOf("Nyakasanza", "Mbunya", "Shari")),
    VilleEntry("Ariwara", "Ituri", emptyList()),
    VilleEntry("Mongwalu", "Ituri", emptyList()),
    VilleEntry("Tshikapa", "Kasaï", listOf("Dibumba I", "Dibumba II", "Kanzala", "Mabondo", "Mbumba")),
    VilleEntry("Kananga", "Kasaï-Central", listOf("Kananga", "Katoka", "Lukonga", "Ndesha", "Nganza")),
    VilleEntry("Tshimbulu", "Kasaï-Central", emptyList()),
    VilleEntry("Mbuji-Mayi", "Kasaï-Oriental", listOf("Bipemba", "Dibindi", "Diulu", "Kanshi", "Muya")),
    VilleEntry("Boma", "Kongo-Central", listOf("Kabondo", "Kalamu", "Nzadi")),
    VilleEntry("Matadi", "Kongo-Central", listOf("Matadi", "Mvuzi", "Nzanza")),
    VilleEntry("Mbanza-Ngungu", "Kongo-Central", emptyList()),
    VilleEntry("Inkisi", "Kongo-Central", emptyList()),
    VilleEntry("Kenge", "Kwango", listOf("Cinq-Mai", "Laurent-Désiré-Kabila", "Manonga", "Masikita", "Mavula")),
    VilleEntry("Bandundu", "Kwilu", listOf("Basoko", "Disasi", "Mayoyo")),
    VilleEntry("Kikwit", "Kwilu", listOf("Kazamba", "Kikwit", "Lukemi", "Lukolela", "Nzinda")),
    VilleEntry("Kabinda", "Lomami", listOf("Kabondo", "Kabuela-Buela", "Kajiba", "Mudingayi")),
    VilleEntry("Mwene-Ditu", "Lomami", listOf("Bondoyi", "Musadi", "Mwene-Ditu")),
    VilleEntry("Kolwezi", "Lualaba", listOf("Dilala", "Kolwezi", "Manika")),
    VilleEntry("Kasaji", "Lualaba", emptyList()),
    VilleEntry("Inongo", "Mai-Ndombe", listOf("Bonse", "Inongo", "Mpolo", "Mpongonzoli")),
    VilleEntry("Nioki", "Mai-Ndombe", emptyList()),
    VilleEntry("Kindu", "Maniema", listOf("Alunguli", "Kasuku", "Kindu", "Mikelenge")),
    VilleEntry("Kalima", "Maniema", emptyList()),
    VilleEntry("Lisala", "Mongala", listOf("Bolikango", "Lisala", "Mongala")),
    VilleEntry("Bumba", "Mongala", emptyList()),
    VilleEntry("Gbadolite", "Nord-Ubangi", listOf("Gbadolite", "Molegbe", "Nganza")),
    VilleEntry("Beni", "Nord-Kivu", listOf("Beu", "Bungulu", "Mulekera", "Ruwenzori")),
    VilleEntry("Butembo", "Nord-Kivu", listOf("Bulengera", "Butembo", "Kimemi", "Mususa", "Vulamba")),
    VilleEntry("Goma", "Nord-Kivu", listOf("Goma", "Karisimbi")),
    VilleEntry("Lusambo", "Sankuru", listOf("Kabondo", "Lupembe", "Lusambo", "Tusuanganyi")),
    VilleEntry("Wembo-Nyama", "Sankuru", listOf("Ewango", "Lumumbaville", "Wembo-Nyama")),
    VilleEntry("Bukavu", "Sud-Kivu", listOf("Bagira", "Ibanda", "Kadutu")),
    VilleEntry("Baraka", "Sud-Kivu", emptyList()),
    VilleEntry("Kamituga", "Sud-Kivu", emptyList()),
    VilleEntry("Uvira", "Sud-Kivu", emptyList()),
    VilleEntry("Gemena", "Sud-Ubangi", listOf("Gbazubu", "Labo", "Lac-Ntumba", "Mont-Gila")),
    VilleEntry("Zongo", "Sud-Ubangi", emptyList()),
    VilleEntry("Kalemie", "Tanganyika", listOf("Kalemie", "Lac", "Lukuga")),
    VilleEntry("Kisangani", "Tshopo", listOf("Kabondo", "Kisangani", "Lubunga", "Makiso", "Mangobo", "Tshopo")),
    VilleEntry("Yangambi", "Tshopo", emptyList()),
    VilleEntry("Boende", "Tshuapa", listOf("Boende", "Tshuapa")),
).sortedBy { it.name }

fun communesForVille(ville: String): List<String> =
    DRC_VILLES.find { it.name.equals(ville, ignoreCase = true) }?.communes ?: emptyList()

// Les 26 provinces de la RDC — champ "Province" du formulaire, placé avant la ville.
val DRC_PROVINCES: List<String> = listOf(
    "Bas-Uele", "Équateur", "Haut-Katanga", "Haut-Lomami", "Haut-Uele", "Ituri", "Kasaï",
    "Kasaï-Central", "Kasaï-Oriental", "Kinshasa", "Kongo-Central", "Kwango", "Kwilu", "Lomami",
    "Lualaba", "Mai-Ndombe", "Maniema", "Mongala", "Nord-Kivu", "Nord-Ubangi", "Sankuru",
    "Sud-Kivu", "Sud-Ubangi", "Tanganyika", "Tshopo", "Tshuapa",
)

/** Province d'une ville de la liste, ou null pour une ville saisie librement. */
fun provinceForVille(ville: String): String? =
    DRC_VILLES.find { it.name.equals(ville, ignoreCase = true) }?.province

/** Villes proposées pour une province (toutes si aucune province n'est choisie). */
fun villesForProvince(province: String): List<VilleEntry> =
    if (province.isBlank()) DRC_VILLES
    else DRC_VILLES.filter { it.province.equals(province, ignoreCase = true) }
